File: server/webapp/WEB-INF/rails.new/app/presenters/api_v2/dashboard/modification_representer.rb
##########################################################################
# Copyright 2017 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV2
  module Dashboard
    class ModificationRepresenter < ApiV2::BaseRepresenter
      alias_method :modification, :represented

      def initialize(options)
        @material = options[:material]
        @latest_revision = options[:latest_revision]

        super(options[:modification])
      end

      link :vsm do |opts|
        opts[:url_builder].vsm_show_material_url({material_fingerprint: @material.getFingerprint(), revision: modification.revision})
      end

      property :userName, as: :user_name, skip_nil: true
      property :emailAddress, as: :email_address, skip_nil: true
      property :revision, skip_nil: true
      property :modifiedTime, as: :modified_time, skip_nil: true
      property :comment, skip_nil: true

    end
  end
end

File: api/api-dashboard-v2/src/main/java/com/thoughtworks/go/apiv2/dashboard/representers/ModificationRepresenter.java
/*
 * Copyright 2018 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv2.dashboard.representers;

import com.google.common.collect.ImmutableMap;
import com.thoughtworks.go.api.representers.JsonWriter;
import com.thoughtworks.go.domain.materials.Material;
import com.thoughtworks.go.domain.materials.Modification;
import com.thoughtworks.go.spark.RequestContext;

import java.util.Map;

public class ModificationRepresenter {

    private static final String VSM_HREF = "/materials/value_stream_map/${material_fingerprint}/${revision}";

    public static Map toJSON(Modification model, RequestContext requestContext, Material material) {
        return new JsonWriter(requestContext)

                .addLink("vsm", VSM_HREF, ImmutableMap.of(
                        "material_fingerprint", material.getFingerprint(),
                        "revision", model.getRevision()))

                .addIfNotNull("user_name", model.getUserName())
                .addIfNotNull("email_address", model.getEmailAddress())
                .addIfNotNull("revision", model.getRevision())
                .addIfNotNull("modified_time", model.getModifiedTime())
                .addIfNotNull("comment", model.getComment())
                .getAsMap();
    }

}
